File: Software/Protocol_pico/src/graph.cpp
#include "graph.h"

namespace LannoLeaf {

  Graph::Graph() { }
  
  Graph::~Graph() { 
    std::map<uint8_t, Node *>::iterator i;
    for (i = map.begin(); i != map.end(); i++) {
      delete i->second;
    }
  }

  void Graph::add_node(const uint8_t &address) {
    nmap::iterator itr = map.find(address);

    if (itr == map.end()) {
      Node *node;
      node = new Node(address);
      map[address] = node;
      return;
    }
  }

  void Graph::add_edge(const uint8_t& from, const uint8_t& to) {
    Node *f = (map.find(from)->second);
    Node *t = (map.find(to)->second);
    f->adj.push_back(t);
    t->adj.push_back(f);
  }

  void Graph::print(void) {
    std::map<uint8_t, Node *>::iterator i;
    for (i = map.begin(); i != map.end(); i++) {
      printf("Node: 0x%02x\r\n", i->second->i2c_address);

      printf("Adjecent nodes: ");
      for (Node * adj_node : i->second->adj) {
        printf("0x%02x, ", adj_node->i2c_address);
      }

      printf("\r\n");
    }
  }

}

File: software/LannooleafSpiController/src/graph.js
class Node {
  constructor(address) {
    this.address = address;
    this.adj = new Map();
  }
}

export default class Graph {

  constructor() {
    this.map = new Map();
  }

  AddNode(address) {
    if (!this.map.has(address)) {
      this.map.set(address, new Node(address));
    }
  }

  AddEdge(from, from_side, to) {
    if (this.map.has(from) && this.map.has(to)) {
      let node_from = this.map.get(from);
      let node_to = this.map.get(to);

      node_from.adj.set(from_side, node_to);
    }
  }
  
  Get() {
    return this.map.get(0xff);
  }

}